File: ImplementQueueUsingStacks.py
# Time Complexity: O(1) for push(), amortized O(1) for pop() and peek(), O(1) for empty().
# Space Complexity: O(n), where n is the total number of elements in the queue.
# Approach: Used two stacks (inStack and outStack) to implement FIFO behavior of a queue.

class MyQueue:

    def __init__(self):
        self.inStack = []  # Stack for pushing new elements
        self.outStack = []  # Stack for popping/peeking elements

    def push(self, x: int) -> None:
        self.inStack.append(x)  # Push to inStack

    def pop(self) -> int:
        self.peek()  # Ensure outStack is populated
        return self.outStack.pop()  # Pop from outStack

    def peek(self) -> int:
        if not self.outStack:  # If outStack is empty, transfer elements
            while self.inStack:
                self.outStack.append(self.inStack.pop())
        return self.outStack[-1]  # Return the top element of outStack

    def empty(self) -> bool:
        # Queue is empty if both stacks are empty
        return not self.inStack and not self.outStack



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Implement Queue using Stacks.java
//using 2 stacks instack will work for pushing th ele, out stack will work for popping the elements


class MyQueue {
    Stack<Integer> inSt;
     Stack<Integer> outSt;
    public MyQueue() {
 
     this.inSt = new Stack<>();
     this.outSt = new Stack<>();

    }
    
    public void push(int x) {//tc-O(1)
         inSt.push(x);
    }
    
    public int pop() {//avg case O(1), only few cases it have to insert element first 
    //then pop in that case it will be O(n), 
        if(outSt.isEmpty()){
            while(!inSt.isEmpty()){
                outSt.push(inSt.pop());
            }
        }
        return outSt.pop();
    }
    
    
    public int peek() {//avg case O(1), only few cases it have to insert element first 
    //then peep in that case it will be O(n)
        if(outSt.isEmpty()){
            while(!inSt.isEmpty()){
                outSt.push(inSt.pop());
            }
        }
        return outSt.peek();  
    }
    
    public boolean empty() {//isEmpty will not take any extra time, under the hood
        return inSt.isEmpty() && outSt.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */